frontend/public/components/network-policies/network-policy-model.ts

```diff
@@ -0,0 +1,175 @@
+import * as _ from 'lodash';
+import {
+  NetworkPolicyKind,
+  NetworkPolicyPort as K8SPort,
+  NetworkPolicyPeer as K8SPeer,
+  Selector,
+} from '../../module/k8s';
+
+// Reference: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#networkpolicyspec-v1-networking-k8s-io
+
+export interface NetworkPolicy {
+  name: string;
+  namespace: string;
+  podSelector: string[][];
+  ingress: NetworkPolicyRules;
+  egress: NetworkPolicyRules;
+}
+
+export interface NetworkPolicyRules {
+  rules: NetworkPolicyRule[];
+  denyAll: boolean;
+}
+
+export interface NetworkPolicyRule {
+  key: string;
+  peers: NetworkPolicyPeer[];
+  ports: NetworkPolicyPort[];
+}
+
+export interface NetworkPolicyPeer {
+  key: string;
+  podSelector?: string[][];
+  namespaceSelector?: string[][];
+  ipBlock?: NetworkPolicyIPBlock;
+}
+
+export interface NetworkPolicyIPBlock {
+  cidr: string;
+  except: string[];
+}
+
+export type NetworkPolicyPort = {
+  protocol: string;
+  port: string;
+};
+
+const networkPolicyTypeIngress = 'Ingress';
+const networkPolicyTypeEgress = 'Egress';
+
+interface ConversionError {
+  kind: 'invalid' | 'unsupported';
+  error: string;
+}
+
+const isError = <T>(result: T | ConversionError): result is ConversionError => {
+  return result && (result as ConversionError).error !== undefined;
+};
+export const isNetworkPolicyConversionError = isError;
+
+const factorOutError = <T>(list: (T | ConversionError)[]): T[] | ConversionError => {
+  const err = list.find((r) => isError(r)) as ConversionError | undefined;
+  if (err) {
+    return err;
+  }
+  return list as T[];
+};
+
+const selectorToK8s = <T>(
+  selector: string[][],
+  emptyValue: Selector | undefined,
+): Selector | undefined | ConversionError => {
+  const filtered = selector.filter((pair) => pair.length >= 2 && pair[0] !== '');
+  if (filtered.length > 0) {
+    const obj = _.fromPairs(filtered);
+    if (Object.keys(obj).length !== filtered.length) {
+      return { kind: 'invalid', error: 'Duplicate keys found in label selector' };
+    }
+    return {
+      matchLabels: obj,
+    };
+  }
+  return emptyValue;
+};
+
+type Rule = { from?: K8SPeer[]; to?: K8SPeer[]; ports?: K8SPort[] };
+
+const ruleToK8s = (
```

Comment:
Yes I thought about it :-) there are pros and cons.
True, it's a lot of code that could be just wiped out.
But I don't think it's more error prone, quite the opposite, because the k8s model is not always very consistent, and I don't like to see code to deal with inconsistencies leaking in all the various components, I prefer to define "clean" types and have a more solid base to work with in the underlying components.

To give some examples of the inconsistencies:
- The way a policy affects "ingress" or "egress" rules are different depending on presence of fields and content of policyType field (probably for legacy reasons). In the rewritten model, they obey the same rule.
- Selector field being present or not, empty or not doesn't have exactly the same consequence depending on where it is in the resource definition (parent or peer).

It's also easier to cover how these inconsistencies are handled in tests by calling pure / stateless typescript functions rather than through react component introspection.

And last point, the rewritten model can be adapted to better match the form structure, e.g. a `denyAll` bool to map to checkboxes, instead of a playing with the combination of `policyType` and `ingress` / `egress` fields.

So to sum up, even if it adds hundred lines of code, I think it makes the react component code easier to read and to reason about in the end.